File: Solver.py
import random


def main():
    # random grid generation driver code
    grid = [[0] * 9 for i in range(9)]
    for index, row in enumerate(grid):
        for i in range(len(row)):
            if not random.randrange(0, 4):
                temp = random.randrange(0, 10)
                if isValid(grid, index, i, temp):
                    row[i] = temp
        print(row)
    print("\n")

    if solve(grid):
        for row in grid:
            print(row)
    else:
        print("No Solution")

# main recursive backtracking function
# traverses depth first, backtracking if no valid solutions are possible on current branch
def solve(grid):
    row, col, flag = getNext(grid)
    if not flag:
        return True # grid is solved
    else:
        for i in range(1, len(grid) + 1):
            if isValid(grid, row, col, i):
                grid[row][col] = i
                if solve(grid):
                    return True
                else:
                    grid[row][col] = 0
    return False # no solutions exist

# finds next empty space in the grid
def getNext(grid):
    for i in range(len(grid)):
        for j in range(len(grid)):
            if grid[i][j] == 0:
                return i, j, 1
    return 0, 0, 0


# calls functions to determine if a given number placement is valid
def isValid(grid, row, col, num):
    if grid[row][col]:
        return False
    else:
        return checkRow(grid, row, num) and checkCol(grid, col, num) and checkSquare(grid, row, col, num)

#checks for valid placement incurrent row
def checkRow(grid, row, num):
    for i in grid[row]:
        if num == i: return False
    return True

# checks for valid placement in current column
def checkCol(grid, col, num):
    for i in range(len(grid)):
        if num == grid[i][col]: return False
    return True

# checks for valid placement in current square area
def checkSquare(grid, row, col, num):
    rowStart = int((row // (len(grid) ** 0.5)) * (len(grid) ** 0.5))
    colStart = int((col // (len(grid) ** 0.5)) * (len(grid) ** 0.5))
    for i in range(rowStart, rowStart + int((len(grid) ** 0.5))):
        for j in range(colStart, colStart + int((len(grid) ** 0.5))):
            if num == grid[i][j]: return False
    return True

if __name__ == '__main__':
    main()

File: solver.cpp
#include <iostream>
#include <cmath>
#include <cstdlib>
#include <fstream>
#include <iomanip>
#include <vector>
#include <algorithm>

using namespace std;
class Sudoku {
 public:

 	//function to accept grid input from file
	void input(istream &in) {
		for (int i = 0; i < MAX; i++) {
			for (int j = 0; j < MAX; j++) {
				in >> grid[i][j];
				edit[i][j] = !(grid[i][j]);
			}
		}
	}

	//function to output grid
	void output(ostream &out) {
		for (int i = 0; i < MAX; i++) {
			for (int j = 0; j < MAX; j++) {
				out << grid[i][j] << " ";
				if ((j + 1) % 3 == 0) {
					out << " ";
				}
			}
			out << endl;
			if ((i + 1) % 3 == 0) {
				out << endl;
			}
		}
	}

	//main recursive backtracking function
	//traverses depth first, backtracking if no valid solutions are possible on current branch
	bool solve() {
		int row = 0, col = 0;
		if (!nextEmpty(row, col)) {
			return true; //a solution has been found
		}

		for (int i = 1; i < MAX + 1; i++) {
			if (isValid(i, row, col)) {
				grid[row][col] = i;
				if (solve()) {
					return 1;
				} else {
					grid[row][col] = 0;
				}
			}
		}
		return false; //a solution is impossible
	}

	//public function to call placement checking helper functions
	bool isValid(int num, int row, int col) {
		if (grid[row][col]) { //if spot has already been assigned
			return 0;
		}
		return (checkRow(num, row) && checkCol(num, col) && checkSquare(num, row, col));
	}

 private:
	static const int MAX = 9;
	int grid[MAX][MAX];
	int edit[MAX][MAX];

	//determines if placement is valid in current row
	bool checkRow(int num, int row) {
		for (int i = 0; i < MAX; i++) {
			if (grid[row][i] == num) {
				return false;
			}
		}
		return true;
	}

	//determines if placement is valid in current column
	bool checkCol(int num, int col) {
		for (int i = 0; i < MAX; i++) {
			if (grid[i][col] == num) {
				return false;
			}
		}
		return true;
	}

	//determines if placement is valid in current grid square
	bool checkSquare(int num, int row, int col) {
		int rowStart = (row < 3 ? 0 : row < 6 ? 3 : 6);
		int colStart = (col < 3 ? 0 : col < 6 ? 3 : 6);
		for (int i = rowStart; i < rowStart + 3; i++) {
			for (int j = colStart; j < colStart + 3; j++) {
				if (grid[i][j] == num) {
					return false;
				}
			}
		}
		return true;
	}

	//finds next empty space in grid
	bool nextEmpty(int &row, int &col) {
		for (int i = 0; i < MAX; i++) {
			for (int j = 0; j < MAX; j++) {
				if (!grid[i][j]) {
					row = i;
					col = j;
					return 1;
				}
			}
		}
		return 0;
	}
};

//driver code
int main() {
	ifstream fin("grid.txt");
	Sudoku obj;
	obj.input(fin);
	obj.output(cout);
	cout << endl;
	obj.solve();
	obj.output(cout);
	cout << endl;

	return EXIT_SUCCESS;
}

